### PR TITLE
Make more of multiple dispatch and remove unnecessary let

### PR DESCRIPTION
**Pull request by Eggiverse**
My pull request consists of ~~three~~ two parts:

1. Replace `if` block with **multiple dispatch** which is usually preferred in Julia
~~2. The `let` blocks seem unnecessary to me. Please let me know if it is not the case.~~
3. It is recommended to use `===` rather than `==` when comparing with `nothing` .

### CODE REVIEW

File: src/hamiltonian.jl [hunk 1]
```diff
@@ -7,38 +7,34 @@ end
...
+        generic_derivative(q0, v->H(v, x, p), v)
+    end
+    return ODEProblem(DynamicalODEFunction{false}(dp,dq), ArrayPartition(p0,q0), tspan, p; kwargs...)
+end
+
+function HamiltonianProblem{true}(H,p0,q0,tspan,p=nothing;kwargs...)
+    cfg2 = ForwardDiff.GradientConfig(PhysicsTag(), q0)
+    dp = function (dv,v,x,p,t)
+        fun2 = x->-H(v, x, p)
+        ForwardDiff.gradient!(dv, fun2, x, cfg2, Val{false}())
```
Comment (ChrisRackauckas) [inline src/hamiltonian.jl]: is the closure bug fixed in this case? What's the performance?

**Comment by ChrisRackauckas** [comment]
This is nice if it doesn't lead to the closure performance bug. Last time I checked, it did, but that was v1.0

File: src/hamiltonian.jl [hunk 2]
```diff
@@ -7,38 +7,34 @@ end
...
-function HamiltonianProblem{T}(H,p0,q0,tspan,p=nothing;kwargs...) where T
-    if T == false
-        if typeof(q0) <: Number
-            dp = function (v,x,p,t)
-                ForwardDiff.derivative(x->-H(v, x, p), x)
-            end
-            dq = function (v,x,p,t)
-                ForwardDiff.derivative(v->H(v, x, p), v)
-            end
+function generic_derivative(q0, hami, x)
```
Comment (ChrisRackauckas) [inline src/hamiltonian.jl:10]: this is a very good idea.

File: src/hamiltonian.jl (hunk 1 above)
Comment (Eggiverse) [reply src/hamiltonian.jl]: I haven't tested the performance. I don't understand why `let` could be causing a difference in performance here.
I'll do more tests on this.

Comment (ChrisRackauckas) [reply src/hamiltonian.jl]: It's https://github.com/JuliaLang/julia/issues/15276. It's gotten fixed more and more over the years though.

Comment (Eggiverse) [reply src/hamiltonian.jl]: I think it is better to recover `let` and keep only two other changes since that issue hasn't been fully fixed for now.

Comment (ChrisRackauckas) [reply src/hamiltonian.jl]: sounds good

**Review by ChrisRackauckas — APPROVED** [review]

**Pull request merged**